apache2/apache2_config.c

```diff
@@ -413,10 +413,14 @@ void *merge_directory_configs(apr_pool_t *mp, void *_parent, void *_child)
             ap_log_perror(APLOG_MARK, APLOG_STARTUP|APLOG_NOERRNO, 0, mp, "Using parent rules in this context.");
             #endif
 
-            /* Copy the rules from the parent context. */
-            merged->ruleset = msre_ruleset_create(parent->ruleset->engine, mp);
             /* TODO: copy_rules return code should be taken into consideration. */
-            copy_rules(mp, parent->ruleset, merged->ruleset, child->rule_exceptions);
+            if(child->rule_exceptions->nelts > 0) {
+                /* Copy the rules from the parent context. */
+                merged->ruleset = msre_ruleset_create(parent->ruleset->engine, mp);
```

Comment:
just to make sure I understand, this is about taking care of situation in which there are no child exceptions, right?
otherwise it is the same flow?

Reply:
consider embedding the logic into copy_rules instead of a new function

Reply:
Yes. child->rule_exceptions->nelts > 0 only when we will remove rules at the runtime. In current ruleset, it rarely happens. We want to use msre_ruleset_create_and_copy to improve the performance. If  child->rule_exceptions->nelts > 0, we should go through original path.

Reply:
If we embed the logic into copy_rules, we will waste time to allocate memory to child ruleset during the initialization.

Reply:
i understand, however you can use msre_ruleset_create within msre_ruleset_create_and_copy and avoid replicating the same code

Reply:
I checked the code, msre_ruleset_create will do two things, first allocated for the child ruleset, second, apr_array_make for the phase_request_headers,body...., in the msre_ruleset_create_and_copy , we do need to always allocate memory for the child ruleset, but if parent is NULL, we directly call apr_array_copy for phase_request_headers,body.. If we reuse the msre_ruleset_create, it's not necessary to call apr_array_make if parent is not NULL.